SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityOfferCache.java

```diff
@@ -0,0 +1,160 @@
+        .build();
+
+    pendingDeclines = new HashMap<>();
+  }
+
+  @Override
+  public void cacheOffer(SchedulerDriver driver, long timestamp, Offer offer) {
+    LOG.debug("Caching offer {} for {}", offer.getId().getValue(), JavaUtils.durationFromMillis(configuration.getCacheOffersForMillis()));
+
+    offerCache.put(offer.getId().getValue(), new CachedOffer(offer));
+  }
+
+  @Override
+  public void onRemoval(RemovalNotification<String, CachedOffer> notification) {
+    if (notification.getCause() == RemovalCause.EXPLICIT) {
+      return;
+    }
+
+    LOG.debug("Cache removal for {} due to {}", notification.getKey(), notification.getCause());
+
+    synchronized (offerCache) {
+      if (notification.getValue().offerState == OfferState.AVAILABLE) {
+        declineOffer(notification.getValue());
+      } else {
+        pendingDeclines.put(notification.getKey(), notification.getValue());
+        notification.getValue().expire();
+      }
+    }
+  }
+
+  @Override
+  public void rescindOffer(SchedulerDriver driver, OfferID offerId) {
+    offerCache.invalidate(offerId.getValue());
+  }
+
+  @Override
+  public void useOffer(OfferID offerId) {
+    offerCache.invalidate(offerId.getValue());
+  }
+
+  @Override
+  public List<Offer> checkoutOffers() {
+    List<Offer> offers = new ArrayList<>((int) offerCache.size());
+    for (CachedOffer cachedOffer : offerCache.asMap().values()) {
+      cachedOffer.checkOut();
+      offers.add(cachedOffer.offer);
+    }
+    return offers;
+  }
+
+  @Override
+  public void returnOffer(OfferID offerId) {
+    synchronized (offerCache) {
+      CachedOffer existingCachedOffer = offerCache.getIfPresent(offerId.getValue());
+
+      if (existingCachedOffer == null || existingCachedOffer.offerState == OfferState.EXPIRED) {
+        declineOffer(pendingDeclines.get(offerId.getValue()));
+      } else {
+        existingCachedOffer.checkIn();
+      }
+    }
+  }
+
+  private void declineOffer(CachedOffer offer) {
+    Optional<SchedulerDriver> driver = schedulerDriverSupplier.get();
+
+    if (!driver.isPresent()) {
+      LOG.error("No scheduler driver present to handle expired offer {} - this should never happen", offer.offerId);
+      return;
+    }
+
+    Status status = driver.get().declineOffer(offer.offer.getId());
+
+    LOG.debug("Declined cached offer {} - driver status {}", offer.offerId, status);
+  }
+
+  private enum OfferState {
+    AVAILABLE, CHECKED_OUT, EXPIRED;
+  }
+
+  public static class CachedOffer {
+
+    private final String offerId;
+    private final Offer offer;
+    private OfferState offerState;
+
+    public CachedOffer(Offer offer) {
+      this.offerId = offer.getId().getValue();
+      this.offer = offer;
+      this.offerState = OfferState.AVAILABLE;
+    }
+
+    private void checkOut() {
+      LOG.trace("Checking out offer {}", offerId);
+      Preconditions.checkState(offerState == OfferState.AVAILABLE, "Offer %s was in state %s", offerId, offerState);
```

Comment:
Saw a sentry come through for this the other day. This exception will bubble all the way back up to the LeaderOnlyPoller and cause us to abort. Should we maybe catch this and just skip the offer in the case that something else has it checked out?